dispatcher.go

```diff
@@ -0,0 +1,49 @@
+package inflight
+
+import (
+	"fmt"
+	"sync"
+)
+
+type Dispatcher struct {
+	lock              sync.Mutex
+	queues            []*Queue
+	requestsexecuting int
+	fqScheduler       *FQScheduler
+	ACV               int
+}
+
+func (d *Dispatcher) Run() {
+	// Dispatching is done independently for each priority level. Whenever (1)
+	// a non-exempt priority level's number of running requests is below the
+	// level's assured concurrency value and (2) that priority level has a
+	// non-empty queue, it is time to dispatch another request for service.
+	// The Fair Queuing for Server Requests algorithm below is used to pick a
+	// non-empty queue at that priority level. Then the request at the head of
+	// that queue is dispatched.
+
+	for {
+		func() {
+			d.lock.Lock()
+			defer d.lock.Unlock()
+			// Whenever (1) a non-exempt priority level's number of running
+			// requests is below the level's assured concurrency value
+			if d.requestsexecuting < d.ACV {
+				// and (2) that priority level has a non-empty queue
+				distributionCh, packet := d.fqScheduler.Dequeue()
```

Comment:
i suppose this loop in the dispatcher will cost unnecessary cpu when the server's load is idle. the `Dequeue` should be blocking when all queues are empty until the a next request awakes the loop?

Reply:
This loop will cost unnecessary CPU in all circumstances.  It loops as fast as it can; nowhere in the loop body is waiting for anything other than locking mutexes.

Dispatching will be the result of processing arrivals and completions of requests; we do not need an asynchronous goroutine to do anything about dispatching, the dispatching can be done from the logic handling request arrivals and completions.